trieste/acquisition/function.py

```diff
@@ -519,3 +539,78 @@ def prepare_acquisition_function(
         :param models: The models over each dataset in ``datasets``.
         :return: A batch acquisition function.
         """
+
+
+class BatchReparametrizationSampler:
+    r"""
+    This sampler employs the *reparameterization trick* to approximate batches of samples from a
+    :class:`ProbabilisticModel`\ 's predictive joint distribution as
+
+    .. math:: x \mapsto \mu(x) + \epsilon L(x)
+
+    where :math:`L` is the Cholesky factor s.t. :math:`LL^T` is the covariance, and
+    :math:`\epsilon \sim \mathcal N (0, 1)` is constant for a given sampler, thus ensuring samples
+    form a continuous curve.
+    """
+
+    def __init__(self, sample_size: int, model: ProbabilisticModel):
+        """
+        :param sample_size: The number of samples for each batch of points. Must be positive.
+        :param model: The model to sample from.
+        :raise ValueError (or InvalidArgumentError): If ``sample_size`` is not positive.
+        """
+        tf.debugging.assert_positive(sample_size)
+
+        self._sample_size = sample_size
+
+        # _eps is essentially a lazy constant, declared here but initialised only on the first call
+        # to sample
+        self._eps = tf.Variable(
+            tf.ones([0, 0, sample_size], dtype=tf.float64), shape=[None, None, sample_size]
+        )  # [0, 0, S]
+        self._model = model
+
+    def __repr__(self) -> str:
+        """"""
+        return f"BatchReparametrizationSampler({self._sample_size!r}, {self._model!r})"
+
+    def sample(self, at: TensorType, *, jitter: float = DEFAULTS.JITTER) -> tf.Tensor:
+        """
+        Return approximate samples from the `model` specified at :meth:`__init__`. Multiple calls to
+        :meth:`sample`, for any given :class:`BatchReparametrizationSampler` and ``at``, will
+        produce the exact same samples. Calls to :meth:`sample` on *different*
+        :class:`BatchReparametrizationSampler` instances will produce different samples.
+
+        :param at: Batches of query points at which to sample the predictive distribution, with
+            shape `[..., B, D]`, for batches of size `B` of points of dimension `D`.
+        :param jitter: The size of the jitter to use when stabilising the Cholesky decomposition of
+            the covariance matrix.
+        :return: The samples, of shape `[..., S, B, L]`, where `S` is the `sample_size`, `B` the
+            number of points per batch, and `L` the dimension of the model's predictive
+            distribution.
+        :raise ValueError (or InvalidArgumentError): If ``at`` is a scalar, or ``jitter`` is
+            negative.
+        """
+        tf.debugging.assert_rank_at_least(at, 2)
+        tf.debugging.assert_greater_equal(jitter, 0.0)
+
+        mean, cov = self._model.predict_joint(at)  # [..., B, L], [..., L, B, B]
+        batch_size = at.shape[-2]
+
+        if tf.size(self._eps) == 0:
+            self._eps.assign(
+                tf.random.normal(
```

Comment:
Will this not be wrong if `sample` is called with different batch sizes?

Reply:
yes, that's a real tricky one. Hadn't thought of that at all

Reply:
@vpicheny is it right that we'd only expect someone to use this with a fixed batch size?

Reply:
@vpicheny is it reasonable to require a consistent batch shape for a given sampler (and therefore BO loop)?

Reply:
i'm pretty sure it is, so i'll just make that change

Reply:
yes that seems reasonable

Reply:
@ChrisMorter , fyi victor and I decided to say batch size is fixed for now, and to make a ticket to allow variable batch size (which is probably possible by adding more N(0, 1) or subsampling for bigger/smaller batch sizes)

Reply:
btw the check for the batch size is added above

Reply:
Ok, that seems reasonable.